pkg/cloud/gcp/actuators/machine/reconciler.go

```diff
@@ -195,3 +198,63 @@ func validateMachine(machine machinev1.Machine, providerSpec v1beta1.GCPMachineP
 	// Default values can also be set here
 	return nil
 }
+
+// Returns true if machine exists.
+func (r *Reconciler) instanceExists() (bool, error) {
+	if err := validateMachine(*r.machine, *r.providerSpec); err != nil {
+		return false, fmt.Errorf("failed validating machine provider spec: %v", err)
+	}
+	zone := r.providerSpec.Zone
+	// Need to verify that our project/zone exists before checking machine, as
+	// invalid project/zone produces same 404 error as no machine.
+	if err := r.validateZone(); err != nil {
+		return false, fmt.Errorf("Unable to verify project/zone exists: %v/%v; err: %v", r.projectID, zone, err)
+	}
+	_, err := r.computeService.InstancesGet(r.projectID, zone, r.machine.Name)
+	if err == nil {
+		return true, nil
+	}
+	if isNotFoundError(err) {
+		return false, nil
+	}
+
+	return false, fmt.Errorf("Error getting running instances: %v", err)
+
+}
+
+// Returns true if machine exists.
+func (r *Reconciler) delete() error {
+	exists, err := r.instanceExists()
+	if err != nil {
+		return err
+	}
+	if !exists {
+		klog.Infof("Machine %v not found during delete, skipping", r.machine.Name)
+		return nil
+	}
+	zone := r.providerSpec.Zone
+	operation, err := r.computeService.InstancesDelete(r.projectID, zone, r.machine.Name)
+	if err != nil {
+		return fmt.Errorf("failed to delete instance via compute service: %v", err)
+	}
+	if op, err := r.waitUntilOperationCompleted(zone, operation.Name); err != nil {
+		return fmt.Errorf("failed to wait for delete operation via compute service. Operation status: %v. Error: %v", op, err)
+	}
+	if err != nil {
```

Comment:
Why do we need this line? If calls to `InstanceDelete()` fail we return (line 238). If calls to `waitUntilOperationCompleted()` fail we return (line 241). At this point delete has succeeded.

Reply:
@frobware this is based on upstream implementation and also create.  While the API request my have been successful, it looks like there's some kind of polling mechanism in the background to ensure that the request actually takes place.  Delete and create are async.

Reply:
nvm, I see I left an extra if statement in there.